File: nn/datasets.ts
import { execSync } from 'child_process';
import * as fs from 'fs';

export class DataLoader<T = number[][]> {
  dataset: Dataset<T>;
  batch_size: number;
  shuffle: boolean;
  indices: number[];

  constructor(dataset: Dataset<T>, batch_size = 32, shuffle = true) {
    this.dataset = dataset;
    this.batch_size = batch_size;
    this.shuffle = shuffle;
    this.indices = Array.from({ length: dataset.length() }, (_, i) => i);
    if (shuffle) this._shuffle();
  }

  private _shuffle() {
    for (let i = this.indices.length - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1));
      [this.indices[i], this.indices[j]] = [this.indices[j], this.indices[i]];
    }
  }

  length(): number {
    return Math.ceil(this.dataset.length() / this.batch_size);
  }

  get(batchIdx: number): { xs: T; ys: number[] } {
    const start = batchIdx * this.batch_size;
    const end = Math.min(start + this.batch_size, this.dataset.length());
    const batchIdxArr = this.indices.slice(start, end);
    return this.dataset.get(batchIdxArr);
  }
}

export interface Dataset<T = number[][]> {
  length(): number;
  get(indices: number[]): { xs: T; ys: number[] };
}

export class MnistDataset implements Dataset<number[]> {
  private data: { xs: number[]; ys: number }[] = [];

  constructor(
    private root: string,
    private train = true,
    private transform?: (x: number[]) => number[]
  ) {}

  async load(): Promise<void> {
    const scriptPath = '/tmp/mnist_load.py';
    const outPath = '/tmp/mnist_out.json';
    const pythonScript = `import sys
import json
import os
os.makedirs('${this.root}', exist_ok=True)

try:
    import torch
    from torchvision import datasets, transforms

    transform = transforms.Compose([transforms.ToTensor(), transforms.Normalize((0.5,), (0.5,))])

    if ${this.train ? 'True' : 'False'}:
        ds = datasets.MNIST(root="${this.root}", train=True, download=True, transform=transform)
    else:
        ds = datasets.MNIST(root="${this.root}", train=False, download=True, transform=transform)

    images = []
    labels = []
    for img, label in ds:
        images.append(img.squeeze().tolist())
        labels.append(label)

    with open('${outPath}', 'w') as f:
        json.dump({"images": images, "labels": labels}, f)
    print("DONE")
except Exception as e:
    with open('${outPath}', 'w') as f:
        json.dump({"error": str(e)}, f)
    print("ERROR")
`;
    fs.writeFileSync(scriptPath, pythonScript);
    try {
      execSync(`python3 "${scriptPath}"`, { timeout: 300000 });
      const content = fs.readFileSync(outPath, 'utf-8');
      const result = JSON.parse(content);
      if (result.error) {
        console.error('MNIST load error:', result.error);
        return;
      }
      for (let i = 0; i < result.labels.length; i++) {
        const xs = result.images[i];
        const ys = result.labels[i];
        let transformedXs = xs;
        if (this.transform) {
          transformedXs = this.transform(xs);
        }
        this.data.push({ xs: transformedXs, ys });
      }
    } catch (e) {
      console.error('Failed to load MNIST:', e);
    } finally {
      try { fs.unlinkSync(scriptPath); } catch {}
      try { fs.unlinkSync(outPath); } catch {}
    }
  }

  length(): number {
    return this.data.length;
  }

  get(indices: number[]): { xs: number[]; ys: number[] } {
    const xs: number[] = [];
    const ys: number[] = [];
    for (const idx of indices) {
      const item = this.data[idx];
      xs.push(...item.xs);
      ys.push(item.ys);
    }
    return { xs, ys };
  }
}

export class Compose {
  transforms: ((x: number[]) => number[])[];

  constructor(transforms: ((x: number[]) => number[])[]) {
    this.transforms = transforms;
  }

  __call__(x: number[]): number[] {
    let result = x;
    for (const t of this.transforms) {
      result = t(result);
    }
    return result;
  }
}

export function ToTensor(): (x: number[] | number[][]) => number[] {
  return (x) => (Array.isArray(x[0]) ? (x as number[][]).flat() : x as number[]);
}

export function Normalize(mean: number[], std: number[]): (x: number[]) => number[] {
  return (x: number[]) => x.map((v, i) => (v - mean[i % mean.length]) / std[i % std.length]);
}

export const datasets = {
  MNIST: class {
    private dataset: MnistDataset;

    constructor(root: string, train = true, download = true) {
      const transforms: ((x: number[]) => number[])[] = [
        ToTensor(),
        Normalize([0.5], [0.5]),
      ];
      const composed = (x: number[]) => {
        let result = x;
        for (const t of transforms) {
          result = t(result);
        }
        return result;
      };
      this.dataset = new MnistDataset(root, train, composed);
    }

    async load(): Promise<MnistDataset> {
      await this.dataset.load();
      return this.dataset;
    }
  }
};

File: nn/datasets.rs
//! nn/datasets.rs
//! MNIST dataset loader via Python/torch subprocess.
//! Calls `python -c "..."` to get raw data, parses into Rust Tensor.

use std::process::Command;
use crate::nn::Tensor;

pub struct DataLoader {
    dataset: Dataset,
    batch_size: usize,
    shuffle: bool,
    indices: Vec<usize>,
    pos: usize,
}

pub struct Dataset {
    pub images: Vec<Tensor>,
    pub labels: Vec<usize>,
    pub count: usize,
}

pub fn load_mnist(root: &str) -> Dataset {
    let script = format!(
        r#"
import sys
import pickle
try:
    from torchvision import datasets
    from torchvision.transforms import Compose, ToTensor, Normalize
    transform = Compose([ToTensor(), Normalize((0.1307,), (0.3081,))])
    train_ds = datasets.MNIST('{root}', train=True, download=True, transform=transform)
    test_ds  = datasets.MNIST('{root}', train=False, download=True, transform=transform)
    data = []
    for img, label in train_ds:
        arr = img.squeeze().numpy()
        data.append((arr.tobytes(), arr.shape, label))
    for img, label in test_ds:
        arr = img.squeeze().numpy()
        data.append((arr.tobytes(), arr.shape, label + 10))
    sys.stdout.write(str(len(data)))
except Exception as e:
    sys.stderr.write(str(e))
    sys.exit(1)
"#
    );

    let output = Command::new("python")
        .arg("-c")
        .arg(&script)
        .output()
        .expect("Failed to execute python for MNIST download");

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        eprintln!("MNIST download failed: {}", stderr);
    }

    let count = String::from_utf8_lossy(&output.stdout).trim().parse::<usize>().unwrap_or(0);

    let script2 = format!(
        r#"
import sys
import pickle
import numpy as np
from torchvision import datasets
from torchvision.transforms import Compose, ToTensor, Normalize
transform = Compose([ToTensor(), Normalize((0.1307,), (0.3081,))])
train_ds = datasets.MNIST('{root}', train=True, download=True, transform=transform)
test_ds  = datasets.MNIST('{root}', train=False, download=True, transform=transform)

out = []
for img, label in train_ds:
    arr = img.squeeze().numpy()
    out.append(arr.tobytes().hex() + ',' + str(label) + ',' + str(arr.shape[0]) + ',' + str(arr.shape[1]))
for img, label in test_ds:
    arr = img.squeeze().numpy()
    out.append(arr.tobytes().hex() + ',' + str(label + 10) + ',' + str(arr.shape[0]) + ',' + str(arr.shape[1]))
sys.stdout.write(';'.join(out))
"#
    );

    let output2 = Command::new("python")
        .arg("-c")
        .arg(&script2)
        .output()
        .expect("Failed to execute python for MNIST data");

    let raw = String::from_utf8_lossy(&output2.stdout);
    let mut images = Vec::new();
    let mut labels = Vec::new();

    for entry in raw.split(';') {
        let parts: Vec<&str> = entry.split(',').collect();
        if parts.len() < 4 {
            continue;
        }
        let hex_data = parts[0];
        let label = parts[1].parse::<usize>().unwrap_or(0);
        let h = parts[2].parse::<usize>().unwrap_or(28);
        let w = parts[3].parse::<usize>().unwrap_or(28);

        if let Ok(bytes) = hex::decode(hex_data) {
            let data: Vec<f32> = bytes
                .chunks(4)
                .map(|chunk| {
                    let mut f = [0u8; 4];
                    f.copy_from_slice(chunk);
                    f32::from_le_bytes(f)
                })
                .collect();

            let shape = vec![1, h, w];
            images.push(Tensor::new(data, shape, false));
            labels.push(label);
        }
    }

    let count = images.len();
    Dataset { images, labels, count }
}

impl DataLoader {
    pub fn new(dataset: Dataset, batch_size: usize, shuffle: bool) -> Self {
        let indices: Vec<usize> = (0..dataset.count).collect();
        DataLoader {
            dataset,
            batch_size,
            shuffle,
            indices,
            pos: 0,
        }
    }

    pub fn batch(&mut self) -> Option<(Vec<Tensor>, Vec<usize>)> {
        if self.pos >= self.dataset.count {
            return None;
        }

        if self.pos == 0 && self.shuffle {
            use rand::Rng;
            let mut rng = rand::thread_rng();
            for i in (1..self.indices.len()).rev() {
                let j = rng.gen_range(0..=i);
                self.indices.swap(i, j);
            }
        }

        let start = self.pos;
        let end = (self.pos + self.batch_size).min(self.dataset.count);
        self.pos = end;

        let batch_images: Vec<Tensor> = self.indices[start..end]
            .iter()
            .map(|&i| self.dataset.images[i].clone())
            .collect();
        let batch_labels: Vec<usize> = self.indices[start..end]
            .iter()
            .map(|&i| self.dataset.labels[i])
            .collect();

        Some((batch_images, batch_labels))
    }

    pub fn len(&self) -> usize {
        (self.dataset.count + self.batch_size - 1) / self.batch_size
    }
}

impl Dataset {
    pub fn count(&self) -> usize {
        self.count
    }
}

pub struct Compose {
    transforms: Vec<Box<dyn Transform>>,
}

pub trait Transform: Send + Sync {
    fn transform(&self, img: &[f32], h: usize, w: usize) -> Vec<f32>;
}

pub struct Grayscale;
pub struct Normalize {
    mean: f32,
    std: f32,
}
pub struct RandomRotation {
    _degrees: i32,
}

impl Compose {
    pub fn new(transforms: Vec<Box<dyn Transform>>) -> Self {
        Compose { transforms }
    }
}

impl Transform for Normalize {
    fn transform(&self, img: &[f32], _h: usize, _w: usize) -> Vec<f32> {
        img.iter().map(|x| (x - self.mean) / self.std).collect()
    }
}

impl Transform for Grayscale {
    fn transform(&self, img: &[f32], _h: usize, _w: usize) -> Vec<f32> {
        img.to_vec()
    }
}

impl Transform for RandomRotation {
    fn transform(&self, img: &[f32], h: usize, w: usize) -> Vec<f32> {
        let _size = h * w;
        img.to_vec()
    }
}

impl Compose {
    pub fn apply(&self, img: &[f32], h: usize, w: usize) -> Vec<f32> {
        let mut out = img.to_vec();
        for t in &self.transforms {
            out = t.transform(&out, h, w);
        }
        out
    }
}